enhancements/cguConditionEnhancement.md

```diff
@@ -0,0 +1,452 @@
+
+- Article provided by @ricardo: [Conditions in Kubernetes Controller](https://maelvls.dev/kubernetes-conditions/)
+- [Cert-manager: Automatically provision and manage TLS certificates in Kubernetes](https://github.com/cert-manager/cert-manager) uses the single Ready condition which matches to the current TALM condition type
+- [Kubernetes recommendation on Conditions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)
+- Posted a question in slack channel : #forum-apiserver for any existing design guidelines. Still waiting for some feedback.
+- **Openshift-apiserver, kube-apiserver, machine-config, node-tuning, config-operator** have below conditions:
+
+```yaml
+[root@hv2 ~]# oc get co/openshift-apiserver -oyaml | yq '.status.conditions'
+[
+  {
+    "lastTransitionTime": "2022-03-21T14:38:09Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "False",
+    "type": "Degraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:39:58Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "False",
+    "type": "Progressing"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:43:16Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "True",
+    "type": "Available"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:38:08Z",
+    "message": "All is well",
+    "reason": "AsExpected",
+    "status": "True",
+    "type": "Upgradeable"
+  }
+]
+```
+
+- **Network operator** has below condition:
+
+```yaml
+[root@hv2 ~]# oc get co/network -ojson | jq '.status.conditions'
+[
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "False",
+    "type": "Degraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "False",
+    "type": "ManagementStateDegraded"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:35:06Z",
+    "status": "True",
+    "type": "Upgradeable"
+  },
+  {
+    "lastTransitionTime": "2022-06-07T15:35:25Z",
+    "status": "False",
+    "type": "Progressing"
+  },
+  {
+    "lastTransitionTime": "2022-03-21T14:37:57Z",
+    "status": "True",
+    "type": "Available"
+  }
+]
+```
+
+## Proposal
+
+This section proposes minimalistic changes without breaking the current design pattern. The below proposal/recommendation consists of more granular `status.conditions`  based on all the verification we already have in TALM.
+
+**Considering**: backup,precaching is true and enable for upgrade is true as well. 
+
+### Workflow Description
+
+1. Condition type related to cluster existing in the managedcluster. The type could reflect on the existence of the targeted cluster to be present in the managedcluster. 
+
+  ```yaml
+  Type:    "ClusterValidation/ClustersSelected",
+  Status:  True/False,
+  Reason:  "",
+  Message: “”
+  ```
+
+2. Condition type: **BackupDone**
```

Comment:
For backup and precaching i may change some wording because in another story, we are targeting that, even when backup and precaching is partially done, we should proceed for upgrade to those successful cluster. So, probably BackupDone will become Backup and and the state/reason will be populated in the Reason variable.

Reply:
I think another condition to reflect job creation is better than overload one condition with multiple reasons (the binary status becomes awkward in that case). Pre-caching uses two today. @vitus133 I am curious whether the current PreCachingRequired condition can be aligned with the backup initialized condition as Mohamed proposed.

Reply:
that could work, we will need to set the state and reasons accordingly.

Reply:
@jc-rh I would disagree to the fact that, creating newer conditions for backup/precaching creates lesser values for admin/user. Once backup is started and it failed to launch it for any/various reason it will reflect on its Reason. I would be more inclined to propose the condition name BackupDone/PrecachingDone changed to Backup/Precaching (or any better wording), and keep appropriate reason around it. From the admin/user side, they must not care how the backup is launched and whether it is in initialized state and then proceeded to BackupDone with multiple Reason. In addition, how can we define false status for  backupInitialized since we don't monitor and elevate the status for whether the job was successfully launched or not in the targeted cluster?

Reply:
@sabbir-47 its better view it as two different machine states. By having 2 conditions we can distinguish between initialized state and monitoring backup state. AS we are moving toward not blocking clusters upgrade till all cluster backup done we could have backup initialize and done at same time for different clusters.

Reply:
OR I'm just thinking loud; if we will move backup and precache to be also per batch then having 1 condition will be enough as precache and backup will be initialized and started for all batch clusters at same time.

Reply:
@serngawy Currently, batching is not tied up with backup/precaching. yes, we will block any operation until backup is done/failed in all the clusters, meaning simultaneously backup, precaching and upgrade will not occure. Only thing changes is, after we run through backup to all the clusters, we only will proceed for precaching/upgrade with the successful clusters, leaving the failed ones.

Reply:
I chose backupSucceeded